sql: use a spool operator to fix the correctness of CTE mutations

knz opened the pull request:
Fixes #20732. This is the very strict minimum amount of code needed to fix that correctness issue, without requiring the other changes in #23373. cc @jordanlewis. I'll want to cherry-pick this.

Prior to this patch, if a mutation statement was used in a CTE or as a
data source with [ ... ] it would run the risk to not complete (and
not commit its results properly) if the surrounding query did not
consume its results fully (which can happen e.g. with LIMIT).

This patch fixes that by introducing a spool operator that does the
right thing. The result rows, if any, are accumulated in memory.

Release note (bug fix): fixed a bug where INSERT/DELETE/UPDATE/UPSERT
may lose updates if run using WITH or the [ ... ] syntax.

knz committed:
sql: use a spool operator to fix the correctness of CTE mutations

Prior to this patch, if a mutation statement was used in a CTE or as a
data source with [ ... ] it would run the risk to not complete (and
not commit its results properly) if the surrounding query did not
consume its results fully (which can happen e.g. with LIMIT).

This patch fixes that by introducing a spool operator that does the
right thing. The result rows, if any, are accumulated in memory.

Release note (bug fix): fixed a bug where INSERT/DELETE/UPDATE/UPSERT
may lose updates if run using WITH or the [ ... ] syntax.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23824)
<!-- Reviewable:end -->

knz committed:
sql: avoid accumulating too many rows in the spool if LIMIT is used

This patch prevents the accumulation of rows in memory beyond the
number required by the current LIMIT clause.

Release note: None

knz commented:
Friendly request to look at this later today, so that we can cover any iteration on the code tomorrow and I can issue the cherry-pick PR. I'm not yet 100% sure I'll have time after that on Friday and in the weekend to both iterate and prepare the cherry-pick.

RaduBerinde commented:
This is cool and the code came out very clean.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_limits.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aBP080lXB94kd4p6-:-L7aBP080lXB94kd4p60:blnpnkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e753b9784c9ab6b701023ddca9670b25b4849461/pkg/sql/opt_limits.go#L138)):*
> ```Go
> 			n.hardLimit = numRows
> 		}
> 		p.setUnlimited(n.source)
> ```

I would add a comment here explaining that it's important that we process all the rows in the spool's source.

---

*[pkg/sql/plan.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7a9xMX8WsxbkxSaJge:-L7a9xMX8WsxbkxSaJgf:b-defc51) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/plan.go#L215)):*
> ```Go
> var _ planNodeFastPath = &upsertNode{}
> 
> // planNodeRequireSpool is implemented by nodes whose parent must
> ```

[nit] I would say "serves as a marker for nodes .." since it doesn't actually do anything

---

*[pkg/sql/spool.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aAcUkYntIOaPJIgQ3:-L7aAcUkYntIOaPJIgQ4:bqbg1ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/spool.go#L50)):*
> ```Go
> 		params.EvalContext().Mon.MakeBoundAccount(),
> 		sqlbase.ColTypeInfoFromResCols(planColumns(s.source)),
> 		0,
> ```

/* rowCapacity */

---

*[pkg/sql/spool.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aADRx6WrclEppCGTm:-L7aADRx6WrclEppCGTn:big3ixm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/spool.go#L74)):*
> ```Go
> // FastPathResults implements the planNodeFastPath interface.
> func (s *spoolNode) FastPathResults() (int, bool) {
> 	if f, ok := s.source.(planNodeFastPath); ok {
> ```

this could use some commentary. (The fast path is safe to use without spooling?)

---

*[pkg/sql/spool.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aBsxg1e_SceQX7p0t:-L7aBsxg1e_SceQX7p0u:b68wy59) ([raw file](https://github.com/cockroachdb/cockroach/blob/e753b9784c9ab6b701023ddca9670b25b4849461/pkg/sql/spool.go#L56)):*
> ```Go
> 	)
> 
> 	// Accumulate all the rows. This also guarantees execution of the
> ```

[nit] comment needs updating (we're not accumulating all rows necessarily)

---

*[pkg/sql/logictest/testdata/logic_test/spool, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aAvN-P_y1vF6ndOkn:-L7aAvN-P_y1vF6ndOko:bhgbc43) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/logictest/testdata/logic_test/spool#L152)):*
> ```
> ·            size  1 column, 1 row
> 
> # Check that no spool is inserted for a top-level INSERT, but
> ```

s/inserted/used

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-:-L7aB0e12PaCn3caqtou:b-hmf7ac)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks! Radu I will address your comments tomorrow morning. Jordan that gives you a little time if you want to chime in too.

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> very elegant solution

To be clear, does this also fix #22304?

---

Reviewed 15 of 15 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/spool, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7amKYsiB9ira2FI3S1:-L7amKYsiB9ira2FI3S2:b10ytth) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/logictest/testdata/logic_test/spool#L12)):*
> ```
> ----
> limit                  ·     ·
>  └── spool             ·     ·
> ```

beautiful!

---

*[pkg/sql/logictest/testdata/logic_test/spool, line 164 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7amVuexg23WEjEtxDI:-L7amVuexg23WEjEtxDJ:b111pyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/logictest/testdata/logic_test/spool#L164)):*
> ```
>                 │           into  t(x)
>                 └── values  ·     ·
> ·                           size  1 column, 1 row
> ```

Unless I'm missing something, this is missing an actual execution test. We definitely need one of those, both to exercise the code and as a regression test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-:-L7amyWnLWZ4dHCCVzlg:b-lyn6qz)*
<!-- Sent from Reviewable.io -->

knz commented:
> does this also fix #22304?

No it doesn't. #23373 does this.

---

Review status: 10 of 16 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/opt_limits.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aBP080lXB94kd4p6-:-L7dhNarbNbzWEKgn6l-:b-8unnr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e753b9784c9ab6b701023ddca9670b25b4849461/pkg/sql/opt_limits.go#L138)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would add a comment here explaining that it's important that we process all the rows in the spool's source.
</blockquote></details>

Ha! Actually here the limit can be propagated. Although we _currently_ only use the spool with mutation statements, in general there is no reason why this should be the case. Then it's OK to push the limit through. I have done it here, then added a comment below to remind the reader why delete/insert/update/upsert do not take limits (and should not).

---

*[pkg/sql/plan.go, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7a9xMX8WsxbkxSaJge:-L7di1U4CuxMY3UvdGv5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/plan.go#L215)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I would say "serves as a marker for nodes .." since it doesn't actually do anything
</blockquote></details>

Done.

---

*[pkg/sql/spool.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aAcUkYntIOaPJIgQ3:-L7di56i1twSDNEhymMO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/spool.go#L50)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

/* rowCapacity */
</blockquote></details>

Done.

---

*[pkg/sql/spool.go, line 74 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aADRx6WrclEppCGTm:-L7diwB9FJAbjPBhfEIa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/spool.go#L74)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this could use some commentary. (The fast path is safe to use without spooling?)
</blockquote></details>

Done.

---

*[pkg/sql/spool.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aBsxg1e_SceQX7p0t:-L7dhn2n6i0UP--RVIbV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e753b9784c9ab6b701023ddca9670b25b4849461/pkg/sql/spool.go#L56)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] comment needs updating (we're not accumulating all rows necessarily)
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/spool, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7aAvN-P_y1vF6ndOkn:-L7dj36rKWnlr4gV_WZ8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/logictest/testdata/logic_test/spool#L152)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

s/inserted/used
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/spool, line 164 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-L7amVuexg23WEjEtxDI:-L7dpqVqCLz6-fykhlVd:b-gnuyev) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df6abd408bc01974ae1c902c7dfb482fee84bfd/pkg/sql/logictest/testdata/logic_test/spool#L164)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Unless I'm missing something, this is missing an actual execution test. We definitely need one of those, both to exercise the code and as a regression test.
</blockquote></details>

Done (`statement_source`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-:-L7dptPcUbNklDjxk4O0:bsw71s9)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 10 of 16 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23824#-:-L7drBLmkfq9qstpzwGl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for your speedy but yet sharp reviews!
